motoko/fileupload/src/fileupload/main.mo

```diff
@@ -0,0 +1,165 @@
+                };
+            };
+        };
+
+        return {
+            body = Blob.toArray(Text.encodeUtf8("Permission denied. Could not perform this operation"));
+            headers = [];
+            status_code = 403;
+            streaming_strategy = null;
+        };
+    };
+
+    private func create_strategy(
+        key           : Text,
+        index         : Nat,
+        asset         : Types.Asset,
+        encoding      : Types.AssetEncoding,
+    ) : ?Types.StreamingStrategy {
+        switch (create_token(key, index, encoding)) {
+            case (null) { null };
+            case (? token) {
+                let self: Principal = Principal.fromActor(Assets);
+                let canisterId: Text = Principal.toText(self);
+                let canister = actor (canisterId) : actor { http_request_streaming_callback : shared () -> async () };
+
+                return ?#Callback({
+                    token;
+                    callback = canister.http_request_streaming_callback;
+                });
+            };
+        };
+    };
+
+    public shared query({caller}) func http_request_streaming_callback(
+        st : Types.StreamingCallbackToken,
+    ) : async Types.StreamingCallbackHttpResponse {
+
+        switch (assets.get(st.key)) {
+            case (null) throw Error.reject("key not found: " # st.key);
+            case (? asset) {
+                return {
+                    token = create_token(
+                        st.key,
+                        st.index,
+                        asset.encoding,
+                    );
+                    body = asset.encoding.content_chunks[st.index];
+                };
+            };
+        };
+    };
+
+    private func create_token(
+        key              : Text,
+        chunk_index      : Nat,
+        encoding         : Types.AssetEncoding,
+    ) : ?Types.StreamingCallbackToken {
+
+        if (chunk_index + 1 >= encoding.content_chunks.size()) {
+            null;
+        } else {
+            ?{
+                key;
+                index = chunk_index + 1;
+                content_encoding = "gzip";
+            };
+        };
+    };
+
+    public shared({caller}) func create_chunk(chunk: Types.Chunk) : async {
+        chunk_id : Nat
+    } {
+        nextChunkID := nextChunkID + 1;
+        chunks.put(nextChunkID, chunk);
+
+        return {chunk_id = nextChunkID};
+    };
+
+    public shared({caller}) func commit_batch(
+        {batch_name: Text; chunk_ids: [Nat]; content_type: Text;} : {
+            batch_name: Text;
+            content_type: Text;
+            chunk_ids: [Nat];
+        },
+    ) : async () {
+         var content_chunks : [[Nat8]] = [];
+
+         for (chunk_id in chunk_ids.vals()) {
+            let chunk: ?Types.Chunk = chunks.get(chunk_id);
+
+            switch (chunk) {
+                case (?{content}) {
+                    content_chunks := Array.append<[Nat8]>(content_chunks, [content]);
```

Comment:
```suggestion
                    content_chunks := Array.append<[Nat8]>(content_chunks, [content]);
```
Array.append in a loop is quadratic. Is `chunk_ids just [0,1, ...,n)` for some consecutive indices? In that case, you could use `Array.tabulate<[Nat8]>(n, func i { ...})` to create the outer array without fear of 0(n^2) behaviour.

Reply:
(but I don't fully grok this code, tbh)